### ISSUE TITLE
no response on request without attached cycles

### ISSUE THREAD

**Issue by letmejustputthishere** [dfinity/exchange-rate-canister#228]
it seems like the xrc never responds to a request that is sent without attached cycles and just times out. can we instead return an error with "not enough cycles sent" or smth?

**Comment by THLO** [comment]
That's exactly what the exchange rate canister is supposed to do, see [this test](https://github.com/dfinity/exchange-rate-canister/blob/481c7c65822a63fe830ef77fd5b14bf50bc4d5f0/src/xrc/src/api/test.rs#L214).
You can see all the possible errors [here](https://github.com/dfinity/exchange-rate-canister/blob/481c7c65822a63fe830ef77fd5b14bf50bc4d5f0/src/xrc/xrc.did#L34).

Can you provide instructions on how to reproduce the issue?

**Comment by letmejustputthishere** [comment]
if you call the canister from the command line like this
```
dfx canister --network ic call uf6dk-hyaaa-aaaaq-qaaaq-cai get_exchange_rate '(record { timestamp = null; quote_asset = record {class = variant { Cryptocurrency}; symbol = "BTC"}; base_asset = record {class = variant {Cryptocurrency}; symbol = "USDT"}})'  --candid ~/Downloads/xrc.did
```
the call never returns.

calling it with attached cycles works as expected
```
dfx canister --network ic --wallet obkml-oaaaa-aaaal-abk3q-cai call uf6dk-hyaaa-aaaaq-qaaaq-cai get_exchange_rate '(record { timestamp = null; quote_asset = record {class = variant { Cryptocurrency}; symbol = "BTC"}; base_asset = record {class = variant {Cryptocurrency}; symbol = "USDT"}})' --with-cycles 20000000000 --candid ~/Downloads/xrc.did
```

**Comment by THLO** [comment]
It appears to be a bug in `dfx` 0.14.1: When making an update call to a replica with a replica version that is more recent than the one shipped with `dfx` and the execution results in an error, then `dfx` hangs.

Since the exchange rate canister doesn't cause this behavior, I'll close this issue.

**Comment by letmejustputthishere** [comment]
i had the same issue using candid ui, shouldn't this be a different agent?

**Comment by THLO** [comment]
I just tried it out using the Candid UI canister. When sending 10T cycles, I got an exchange rate back and when sending 10M cycles, I got an error back as expected.

Can you maybe try again? I can't reproduce the issue.

**Comment by letmejustputthishere** [comment]
Out of curiosity, how did you send cycles using Candid UI? I'm using [this deployment](https://a4gq6-oaaaa-aaaab-qaa4q-cai.raw.ic0.app/?id=uf6dk-hyaaa-aaaaq-qaaaq-cai) of Candid UI.
<img width="963" alt="Screenshot 2023-06-27 at 07 58 55" src="https://github.com/dfinity/exchange-rate-canister/assets/32162112/5daec47b-5f69-43b0-8686-5c5d15dda5ee">

**Comment by THLO** [comment]
I actually used the Candid UI canister to interact with a deployed [XRC demo canister](https://github.com/THLO/xrc_demo). In this setup, an error is returned when sending too few cycles.

You're right, the call appears to be hanging for a while when trying to interact with the exchange rate canister directly using the Candid UI canister. I got the following error eventually:

```
Server returned an error:
Code: 400 ()
Body: Specified ingress_expiry not within expected range:
Minimum allowed expiry: 2023-06-27 14:04:57.736939823 UTC
Maximum allowed expiry: 2023-06-27 14:10:27.736939823 UTC
Provided expiry: 2023-06-27 14:04:55.756 UTC
Local replica time: 2023-06-27 14:04:57.736940284 UTC
```
Based on this error, I don't think the exchange rate canister is causing the problem.

**Comment by letmejustputthishere** [comment]
I think that just means that there was no response within the maximum allowed expiry, but if everything went smooth there should be. What makes you think it's not the exchange rate canister causing the problem? The call doesn't return a response for two different agents (rust and js) even though it should. Is there any reason that the response should take a long time if I'm not sending cycles along with the request?  
I tried dfx `0.14.0` and  `0.13.1` with the same result.

**Comment by THLO** [comment]
You're right, I misinterpreted the error message. It looks like the Candid UI canister simply doesn't get a response.
I'll try to get to the bottom of this.

**Comment by letmejustputthishere** [comment]
hey, any news here?

**Comment by THLO** [comment]
Yes! It was indeed a problem not only with `dfx` but also `agent-rs`.
It should work now with `dfx 0.14.2`. An error is also returned now immediately when calling the exchange rate canister from the Candid UI canister.

Could you please check if everything works as expected for you as well?

**Comment by letmejustputthishere** [comment]
The canister rejects both calls via `agent-rs` and `agent-js` now, the response is a bit opaque though. Could this be improved?
```
The replica returned a replica error: Replica Error: reject code CanisterReject, reject message Canister uf6dk-hyaaa-aaaaq-qaaaq-cai rejected the message, error code Some("IC0516")
```
Thanks for looking into it 🙏

**Comment by THLO** [comment]
You're right, the error message could be improved. We will look into it.

I'm closing this issue again because the original problem has been solved.
